lib/Dialect/FIRRTL/Transforms/RemoveUnusedPorts.cpp

```diff
@@ -0,0 +1,172 @@
+
+  for (auto e : llvm::enumerate(ports)) {
+    unsigned index = e.index();
+    auto port = e.value();
+    auto arg = module.getArgument(index);
+
+    // If the port is don't touch or has unprocessed annotations, we cannot
+    // remove the port. Maybe we can allow annotations though.
+    if (hasDontTouch(arg) || !port.annotations.empty())
+      continue;
+
+    // TODO: Handle inout ports.
+    if (port.isInOut())
+      continue;
+
+    // If the port is input and has an user, we cannot remove the
+    // port.
+    if (port.isInput() && !arg.use_empty())
+      continue;
+
+    // If the port is output, then check that the port is only connected to
+    // invalid or constant.
+    if (port.isOutput()) {
+      if (arg.use_empty()) {
+        outputPortConstants.push_back(None);
+      } else if (arg.hasOneUse()) {
+        auto connect = dyn_cast<ConnectOp>(*arg.user_begin());
+        if (!connect || !isa_and_nonnull<InvalidValueOp, ConstantOp>(
+                            connect.src().getDefiningOp()))
+          continue;
+
+        Operation *srcOp;
+        if (auto constant =
+                dyn_cast<ConstantOp>(connect.src().getDefiningOp())) {
+          outputPortConstants.push_back(constant.value());
+          srcOp = constant;
+        } else {
+          assert(isa<InvalidValueOp>(connect.src().getDefiningOp()) &&
+                 "only expect invalid");
+          srcOp = connect.src().getDefiningOp();
+          outputPortConstants.push_back(None);
+        }
+        // Erase connect op because we are going to remove this output ports.
+        connect.erase();
+        if (srcOp->use_empty())
+          srcOp->erase();
+
+      } else {
+        continue;
+      }
+    }
+
+    removalPortIndexes.push_back(index);
+  }
+
+  // If there is nothing to remove, abort.
+  if (removalPortIndexes.empty())
+    return;
+
+  module.erasePorts(removalPortIndexes);
+
+  LLVM_DEBUG(llvm::for_each(removalPortIndexes, [&](unsigned index) {
+               llvm::dbgs() << "Delete port: " << ports[index].name << "\n";
+             }););
+
+  // Rewrite all uses.
+  for (auto *use : instanceGraphNode->uses()) {
+    auto instance = use->getInstance();
+    OpBuilder builder(instance);
+    unsigned outputPortIndex = 0;
+    for (auto index : removalPortIndexes) {
+      auto result = instance.getResult(index);
+      assert(!ports[index].isInOut() && "don't expect inout ports");
+
+      // If the port is input, replace the port with an unwritten wire
+      // so that we can remove use-chains in SV dialect canonicalization.
+      if (ports[index].isInput()) {
+        WireOp wire =
+            builder.create<WireOp>(instance.getLoc(), result.getType());
+        result.replaceUsesWithIf(wire, [&](OpOperand &op) -> bool {
+          // Connects can be deleted directly.
+          if (isa<ConnectOp>(op.getOwner())) {
+            op.getOwner()->erase();
+            return false;
+          }
+          return true;
+        });
```

Comment:
It seems I should have checked whether the inport port is used as a normal wire here.